network/server/tchannelthrift/node/service.go

```diff
@@ -597,46 +613,65 @@ func (s *service) FetchBlocksMetadataRawV2(tctx thrift.Context, req *rpc.FetchBl
 	fetchedMetadata, nextPageToken, err := s.db.FetchBlocksMetadataV2(
 		ctx, nsID, uint32(req.Shard), start, end, req.Limit, req.PageToken, opts)
 	if err != nil {
-		s.metrics.fetchBlocksMetadata.ReportError(s.nowFn().Sub(callStart))
 		return nil, convert.ToRPCError(err)
 	}
+
 	ctx.RegisterFinalizer(resource.FinalizerFn(fetchedMetadata.Close))
 
-	result, err := s.getFetchBlocksMetadataRawV2Result(nextPageToken, opts, fetchedMetadata)
-	// Finalize pooled datastructures regardless of errors because even in the error
-	// case result contains pooled objects that we need to return.
-	ctx.RegisterFinalizer(s.newCloseableMetadataV2Result(result))
+	result, err := s.getFetchBlocksMetadataRawV2Result(ctx, nextPageToken, opts, fetchedMetadata)
 	if err != nil {
-		return nil, err
+		return nil, convert.ToRPCError(err)
 	}
 
+	ctx.RegisterFinalizer(s.newCloseableMetadataV2Result(result))
 	return result, nil
 }
 
 func (s *service) getFetchBlocksMetadataRawV2Result(
+	ctx context.Context,
 	nextPageToken storage.PageToken,
 	opts block.FetchBlocksMetadataOptions,
 	results block.FetchBlocksMetadataResults,
 ) (*rpc.FetchBlocksMetadataRawV2Result_, error) {
+	elements, err := s.getBlocksMetadataV2FromResult(ctx, opts, results)
+	if err != nil {
+		return nil, err
+	}
+
 	result := rpc.NewFetchBlocksMetadataRawV2Result_()
 	result.NextPageToken = nextPageToken
-	result.Elements = s.getBlocksMetadataV2FromResult(opts, results)
+	result.Elements = elements
 	return result, nil
 }
 
 func (s *service) getBlocksMetadataV2FromResult(
+	ctx context.Context,
 	opts block.FetchBlocksMetadataOptions,
 	results block.FetchBlocksMetadataResults,
-) []*rpc.BlockMetadataV2 {
+) ([]*rpc.BlockMetadataV2, error) {
 	blocks := s.pools.blockMetadataV2Slice.Get()
-
 	for _, fetchedMetadata := range results.Results() {
 		fetchedMetadataBlocks := fetchedMetadata.Blocks.Results()
-		id := fetchedMetadata.ID.Data().Bytes()
+
+		var (
+			id          = fetchedMetadata.ID.Bytes()
+			tags        = fetchedMetadata.Tags
+			encodedTags []byte
+		)
+		if tags != nil && tags.Remaining() > 0 {
+			enc := s.pools.tagEncoder.Get()
+			ctx.RegisterFinalizer(enc)
```

Comment:
is the encoder not safe to close immediately after receiving the encoded data?

Reply:
Nope. The encoder still owns the `checked.Bytes` returned

Reply:
Unfortunately not, it just returns a checked bytes ref that it has itself.  I could copy this out, but it seems unnecessary.  

Or I could create a single one and finalize at the end of this method and copy out and put in a bytes pool each time, however that also seems like more work than necessary perhaps.